src/parser/FileParser.kt

```diff
@@ -0,0 +1,108 @@
+     */
+    fun parseCoffeeShopsList(n: Int) {
+        val invalidCoffeeShopsList: MutableList<String> = mutableListOf()
+        if (lines != null) {
+            var lineNumber = 1
+            lines.forEach { it ->
+                val result: List<String> = it.split(",").map { it.trim() }
+                if (isCSVLineValid(result)) {
+                    insertShopIntoList(result[0], result[1], result[2], n)
+                } else {
+                    if (!it.trim().isEmpty()) invalidCoffeeShopsList.add("Line $lineNumber: $it \n")
+                }
+                lineNumber++
+            }
+
+            if (n <= coffeeShopsList.size) {
+                coffeeShopsList.subList(0, n).forEach {
+                    println("${it.shopName}, ${it.shopDistanceToUser}")
+                }
+            } else {
+                if (coffeeShopsList.size > 0) {
+                    coffeeShopsList.forEach { println("${it.shopName}, ${it.shopDistanceToUser}") }
+                    println("\nThe number of CoffeeShops you requested ($n) exceeds the number of valid shops in the .csv file. Only ${coffeeShopsList.size} result(s) printed")
+                } else {
+                    println("There are no valid coffee shops in $fileName to be displayed")
+                }
+            }
+        }
+        // write all invalid lines (if there are any) in a file
+        if (!invalidCoffeeShopsList.isEmpty()) {
+            println("\nAll invalid lines found in $fileName are written in invalid_coffee_shops.txt")
+            File("invalid_coffee_shops.txt").bufferedWriter().use { out ->
+                invalidCoffeeShopsList.forEach {
+                    out.write(it)
+                }
+            }
+        }
+    }
+
+    /**
+     * inserts a CoffeeShop object into the list. The first ${n} elements of the list must be the first ${n} closest shops to the user
+     * and they must be ordered by distance, from the closest to the furthest
+     *
+     * *** coffeeShopName: the name of the coffeeShop
+     * *** coffeeShopCoordinateX: the X coordinate of the coffeeShop
+     * *** coffeeShopCoordinateY: the Y coordinate of the coffeeShop
+     * *** n : the number of elements from the beginning of the list, ordered by their distance to the user
+     */
+    fun insertShopIntoList(coffeeShopName: String, coffeeShopCoordinateX: String, coffeeShopCoordinateY: String, n: Int) {
+        val coffeeShop = CoffeeShop(coffeeShopName, Coordinate(coffeeShopCoordinateX, coffeeShopCoordinateY), Coordinate(userCoordinate.x, userCoordinate.y))
+        var coffeeShopInserted = false
+        // check if the current coffee shop enters the "top n" coffee shops according to their distance and insert it on the right position
+        for (i in 0..(minOf(n, coffeeShopsList.size) - 1)) {
+            if (coffeeShop < coffeeShopsList[i]) {
```

Comment:
This is a interesting approach.

Unfortunately for me it does not make any sens to compare coffee shops, because actually they get compared based on the distance from the current user.

There a couple of things that I think can be improved:
- coffe shop object should not hold the coordonates of the user, it's not its concern
- coffe shop's can be comparable but not to get the distance, I can imagine a dsl like `userCoordonates.distanceTo(coffeShoopCoordonates)`. The distance calculation should be the concern of some other class.